src/orion/algo/dehb/dehb.py

```diff
@@ -0,0 +1,513 @@
+                self.job_infos[self.get_id(new_trial)].append(job_info)
+                self.register_job(job_info)
+
+                # Standard Orion
+                self.register(new_trial)
+                trials.append(new_trial)
+                logger.debug("Suggest new trials %s", new_trial)
+            else:
+                logger.debug("Already suggested %s", new_trial)
+
+                # Do we already have a result for this trial ?
+                result = self.job_results.get(self.get_id(new_trial))
+
+                # Keep track of duplicated jobs per brackets
+                # Bracket is only done after we reach the budget for unique
+                # jobs
+                self.duplicates[str(job_info["budget"])] += 1
+
+                # if so observe it right now and discard
+                if result is not None:
+                    self._dehb_observe(job_info, *result)
+                else:
+                    # else we need to keep track of it to observe it later
+                    self.job_infos[self.get_id(new_trial)].append(job_info)
+
+        return trials
+
+    def observe(self, trials: List[Trial]) -> None:
+        """Observe the `trials` new state of result.
+
+        Parameters
+        ----------
+        trials: list of ``orion.core.worker.trial.Trial``
+           Trials from a `orion.algo.space.Space`.
+
+        """
+        for trial in trials:
+            if not self.has_suggested(trial):
+                logger.debug("Ignore unseen trial %s", trial)
+                continue
+
+            if self.has_observed(trial):
+                logger.debug("Ignore already observed trial %s", trial)
+                continue
+
+            self.register(trial)
+
+            if trial.status == "completed":
+                self.observe_one(trial)
+                logger.debug(
+                    "Observe trial %s (Remaining %d)", trial, len(self.job_infos)
+                )
+
+    def observe_one(self, trial: Trial) -> None:
+        """Observe a single trial"""
+
+        # Get all the job sampled by DEHB, it might be more than one
+        job_infos = self.job_infos.get(self.get_id(trial), [])
+
+        if len(job_infos) == 0:
+            # this should be 100% unreachable because we check
+            # if the trial was suggested inside `observe`
+            logger.error("Could not find trial %s", self.get_id(trial))
+            return
+
+        # Yes, it is odd; fidelity is cost and fitness is objective
+        cost = trial.params[self.fidelity_index]
+        fitness = trial.objective.value
+
+        # Store the result for later, if we sample
+        # a trial that is too alike for us to evaluate
+        results = cost, fitness
+        self.job_results[self.get_id(trial)] = results
+
+        for job_info in job_infos:
+            cost = job_info["budget"]
+            self._dehb_observe(job_info, cost, fitness)
+
+    def _dehb_observe(self, job_info, cost, fitness):
```

Comment:
Nit: add type annotations

Reply:
Factored out in https://github.com/Epistimio/orion/commit/412e0c8fbfcb18883d361cafa39b5292bcef6cc5 and then typed in https://github.com/Epistimio/orion/commit/ebfff61c98c062202355f2ce3c308661e9fc80a1